tooling/nargo/src/ops/foreign_calls.rs

```diff
@@ -134,53 +137,73 @@ impl DefaultForeignCallExecutor {
 
     fn execute_print(foreign_call_inputs: &[ForeignCallParam]) -> Result<(), ForeignCallError> {
         let skip_newline = foreign_call_inputs[0].unwrap_value().is_zero();
-        let display_values: PrintableValueDisplay = foreign_call_inputs
-            .split_first()
-            .ok_or(ForeignCallError::MissingForeignCallInputs)?
-            .1
-            .try_into()?;
-        print!("{display_values}{}", if skip_newline { "" } else { "\n" });
+
+        let foreign_call_inputs =
+            foreign_call_inputs.split_first().ok_or(ForeignCallError::MissingForeignCallInputs)?.1;
+        let display_string = Self::format_printable_value(foreign_call_inputs, skip_newline)?;
+
+        print!("{display_string}");
+
         Ok(())
     }
+
+    fn execute_assert_message(
+        foreign_call_inputs: &[ForeignCallParam],
+    ) -> Result<ACVMForeignCallResult, ForeignCallError> {
+        let display_string = Self::format_printable_value(foreign_call_inputs, true)?;
+        Ok(display_string.into())
+    }
```

Comment:
Using foreign calls to resolve error messages like this makes me a bit sad. The flow for Javascript would then be

1. ACVM hits a failing assertion
2. ACVM boots out a foreign call with the assert message format string.
3. Javascript formats this string and passes it back to the ACVM. (Requires a JS formatting implementation)
4. ACVM then fails, returning the string which Javascript just handed it.

I'd need to think on whether we can do this formatting in the ACVM cleanly (i.e. without having another abstraction leak from Noir into ACVM).

Reply:
Is this the flow where assert messages are not lazy evaluated?

So we instead have:
1. Process an assert message foreign call and save the formatted string.
2. If we hit an assert that fails we use the previously saved message. If we passed the assertion, we clear the assert message.
3. Continue normal execution

So instead we resolve the assert message, then we hit a failing assertion. I previously had the resolved assert message be its own NargoError, but then this can potentially lead to mismatched errors when we process an assert message that has not failed. I can switch to this if you think it makes more sense.

Reply:
Could we possibly simplify this by doing the formatting in Noir? For example, can we carry forward the format string in expect until it reaches assert, then from there expand the string to "print" each part as needed?

E.g.
```rs
fn foo(x: i32) {
    println(f"x is {x}, which should be 3") 
}

// Compiled into
fn foo(x: i32) {
    print("x is");
    print(x);
    println(", which should be 3") 
}
```
Since each format string is given a unique type based on the structure of the arguments it captures, this would probably be possible after monomorphization. Then perhaps we could have some primitives of extending this to `assert` somehow? My thinking is just that ideally formatted strings aren't something that is built into the backend of a language - I think we'd be saving ourselves complexity by trying to compile them out beforehand somehow. I could be missing something though.

Reply:
Going to answer in the main conversation for visibility. Your suggestion is an interesting way to possibly simplify formatted strings, but I do lean towards that it can be done in a followup PR. I would have to think through about how much it simplifies things vs. other options when we start to extend this to `assert`.

Reply:
One issue with splitting this up into multiple `print` calls is that we're then performing many foreign calls for a single println which is not ideal.

Reply:
Good point. Lets move discussion to this issue (https://github.com/noir-lang/noir/issues/4238)